File: medicoscope-site/lib/models/health_event.dart
class HealthEvent {
  final String? id;
  final String patientId;
  final String eventType;
  final Map<String, dynamic> data;
  final String? linkedDoctor;
  final EventEscalation? escalation;
  final DateTime? createdAt;

  HealthEvent({
    this.id,
    required this.patientId,
    required this.eventType,
    this.data = const {},
    this.linkedDoctor,
    this.escalation,
    this.createdAt,
  });

  factory HealthEvent.fromJson(Map<String, dynamic> json) => HealthEvent(
        id: json['_id']?.toString(),
        patientId: json['patientId']?.toString() ?? '',
        eventType: json['eventType'] ?? '',
        data: Map<String, dynamic>.from(json['data'] ?? {}),
        linkedDoctor: json['linkedDoctor']?.toString(),
        escalation: json['escalation'] != null
            ? EventEscalation.fromJson(json['escalation'])
            : null,
        createdAt: json['createdAt'] != null
            ? DateTime.tryParse(json['createdAt'])
            : null,
      );

  Map<String, dynamic> toJson() => {
        'patientId': patientId,
        'eventType': eventType,
        'data': data,
        if (linkedDoctor != null) 'linkedDoctor': linkedDoctor,
        if (escalation != null) 'escalation': escalation!.toJson(),
      };
}

class EventEscalation {
  final String? type;
  final String? status;
  final int? responseTime;

  EventEscalation({this.type, this.status, this.responseTime});

  factory EventEscalation.fromJson(Map<String, dynamic> json) =>
      EventEscalation(
        type: json['type'],
        status: json['status'],
        responseTime: json['responseTime'],
      );

  Map<String, dynamic> toJson() => {
        if (type != null) 'type': type,
        if (status != null) 'status': status,
        if (responseTime != null) 'responseTime': responseTime,
      };
}

File: medicoscope_app/server/src/models/HealthEvent.js
const mongoose = require('mongoose');

const healthEventSchema = new mongoose.Schema({
  patientId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true,
  },
  eventType: {
    type: String,
    enum: ['detection', 'vitals_session', 'chat_session', 'mindspace', 'document_upload', 'triage'],
    required: true,
  },
  data: {
    // Polymorphic — depends on eventType
    // detection: { className, confidence, category, explanation }
    // vitals_session: { duration, avgHR, avgBP, alerts[] }
    // chat_session: { sessionId, topicSummary }
    // mindspace: { transcript, urgency, riskLevel }
    // document_upload: { documentType, keyFindings[], urgency }
    // triage: { recommendedModule, urgency, reasoning }
    type: mongoose.Schema.Types.Mixed,
    default: {},
  },
  linkedDoctor: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    default: null,
  },
  escalation: {
    type: { type: String, enum: ['immediate', 'priority', 'routine', null], default: null },
    status: { type: String, enum: ['pending', 'acknowledged', 'resolved', null], default: null },
    responseTime: { type: Number, default: null }, // minutes
  },
}, {
  timestamps: true,
});

healthEventSchema.index({ patientId: 1, createdAt: -1 });
healthEventSchema.index({ patientId: 1, eventType: 1 });

module.exports = mongoose.model('HealthEvent', healthEventSchema);
